torch/jit/__init__.py

```diff
@@ -1573,115 +1572,94 @@ def __getstate__(self):
-                    constants_set.add(name)
-
-            # Copy constants
-            self.__dict__["_constants_set"] = constants_set
-            for name in self.__dict__["_constants_set"]:
-                if hasattr(original, name):
-                    if (name in original._parameters or name in original._buffers) and item is not None:
-                        # for 'None' parameters/buffers, don't actually add their values if it exists
-                        continue
-                    ScriptModule.__setattr__(self, name, getattr(original, name))
-
-            # Copy overloads
-            self.__dict__["_overloads"] = dict(getattr(original, "__overloads__", {}))
-
-            self.__dict__["_initialized"] = True
-            _create_methods_from_stubs(self, stubs)
-
-        def __getattr__(self, attr):
-            # Try to get the attribute directly, if that fails, fall back to the
-            # weak module itself
-            try:
-                return ScriptModule.__getattr__(self, attr)
-            except AttributeError as e:
-                # unwrap the original
-                original_module = self.__dict__["_original"]()
-                if original_module and self.__dict__["_initialized"]:
-                    # get attr from original if it is still alive
-                    return getattr(original_module, attr)
-
-                # If it's not on this module and it wasn't on the original
-                # module (or the original is dead), throw the exception
-                raise e
-
-        def __setattr__(self, attr, value):
-            # Once constructed, no new properties can be set
-
-            if not self.__dict__["_initialized"]:
-                # If constructing, don't fall back to original module
-                return ScriptModule.__setattr__(self, attr, value)
-
-            if hasattr(self, attr):
-                return ScriptModule.__setattr__(self, attr, value)
-            else:
-                raise AttributeError("Cannot set new attribute '{}' on "
-                                     "weak script module once it has been "
-                                     "created".format(attr))
-
 else:
     class ScriptModule(torch.nn.Module):
         def __init__(self, optimize=True):
             super(ScriptModule, self).__init__()
 
 
+def _copy_module_to_script_module(module):
+    """
+    This copies Parameters, buffers, submodules, attributes, and constants from
+    an nn.Module into a fresh ScriptModule
+    """
+    script_module = ScriptModule()
+
+    # A quick check that the module's super() is called so everything is set up
+    # for this method to do the copying it needs to
+    if not hasattr(module, '_parameters'):
+        raise RuntimeError("'{}' has not been initialized, did you forget to call 'super()'?"
+                           .format(type(module).__name__))
+
+
+    # HACK: To support None nn.Parameters and buffers, don't copying them if they
```

Comment:
I'm finding our handling of None-type params/buffers really confusing. I assume this is all because we expect params/buffers to be of type `Tensor` rather than `Optional[Tensor]`. So basically we're saying "you can have a None parameter, but *only if* you tell us in `__constants__` that it will be None forever?

Reply:
Pretty much, it'd be really tedious to have to unwrap `Tensor`s all the time, see #14715 for more context